packages/host/cpp/RuntimeNodeApiAsync.cpp

```diff
@@ -0,0 +1,183 @@
+#include "RuntimeNodeApiAsync.hpp"
+#include <ReactCommon/CallInvoker.h>
+#include "Logger.hpp"
+
+struct AsyncJob {
+  using IdType = uint64_t;
+  enum State { Created, Queued, Completed, Cancelled, Deleted };
+
+  IdType id{};
+  State state{};
+  napi_env env;
+  napi_value async_resource;
+  napi_value async_resource_name;
+  napi_async_execute_callback execute;
+  napi_async_complete_callback complete;
+  void* data{nullptr};
+
+  static AsyncJob* fromWork(napi_async_work work) {
+    return reinterpret_cast<AsyncJob*>(work);
+  }
+  static napi_async_work toWork(AsyncJob* job) {
+    return reinterpret_cast<napi_async_work>(job);
+  }
+};
+
+class AsyncWorkRegistry {
+ public:
+  using IdType = AsyncJob::IdType;
+
+  std::shared_ptr<AsyncJob> create(napi_env env,
+      napi_value async_resource,
+      napi_value async_resource_name,
+      napi_async_execute_callback execute,
+      napi_async_complete_callback complete,
+      void* data) {
+    const auto job = std::shared_ptr<AsyncJob>(new AsyncJob{
+        .id = next_id(),
+        .state = AsyncJob::State::Created,
+        .env = env,
+        .async_resource = async_resource,
+        .async_resource_name = async_resource_name,
+        .execute = execute,
+        .complete = complete,
+        .data = data,
+    });
+
+    jobs_[job->id] = job;
+    return job;
+  }
+
+  std::shared_ptr<AsyncJob> get(napi_async_work work) const {
+    const auto job = AsyncJob::fromWork(work);
+    if (!job) {
+      return {};
+    }
+    if (const auto it = jobs_.find(job->id); it != jobs_.end()) {
+      return it->second;
+    }
+    return {};
+  }
+
+  bool release(IdType id) {
+    if (const auto it = jobs_.find(id); it != jobs_.end()) {
+      it->second->state = AsyncJob::State::Deleted;
+      jobs_.erase(it);
+      return true;
+    }
+    return false;
+  }
+
+ private:
+  IdType next_id() {
+    if (current_id_ == std::numeric_limits<IdType>::max()) [[unlikely]] {
+      current_id_ = 0;
+    }
+    return ++current_id_;
+  }
+
+  IdType current_id_{0};
+  std::unordered_map<IdType, std::shared_ptr<AsyncJob>> jobs_;
+};
+
+static std::weak_ptr<facebook::react::CallInvoker> callInvoker;
```

Comment:
While this likely works in most cases, there could be multiple call invokers in flight, especially during hot-module reloading where the JSI runtime is being torn down and recreated for the "new" app instance.

My instinct would be to store the call invoker in a structure which would be indexable by the `napi_env` to ensure the correct call invoker is used for the correct `napi_env`.

Reply:
I would live for this to be addressed in this PR, if you agree and see a way forward. Alternatively, we could merge and release this and create an issue to track an improvement to make it more robust.